// inversify/es/inversify.js
/**
 * inversify v.2.0.0-beta.5 - A lightweight IoC container written in TypeScript.
 * Copyright (c) 2015 Remo H. Jansen
 * MIT inversify.io/LICENSE
 * http://inversify.io
 */
import Kernel from "./kernel/kernel";
import injectable from "./annotation/injectable";
import tagged from "./annotation/tagged";
import named from "./annotation/named";
import inject from "./annotation/inject";
import { makePropertyInjectDecorator, makePropertyMultiInjectDecorator } from "./annotation/property_injectors";
import multiInject from "./annotation/multi_inject";
import targetName from "./annotation/target_name";
import { decorate } from "./annotation/decorator_utils";
import { traverseAncerstors, taggedConstraint, namedConstraint, typeConstraint } from "./syntax/constraint_helpers";
export { Kernel };
export { decorate };
export { injectable };
export { tagged };
export { named };
export { inject };
export { makePropertyInjectDecorator, makePropertyMultiInjectDecorator };
export { multiInject };
export { targetName };
export { traverseAncerstors };
export { taggedConstraint };
export { namedConstraint };
export { typeConstraint };
